### PR TITLE
[HIPPEROS] Implementation of mutex types

### PR DESCRIPTION
**Pull request by odesenfans**
Implemented the pthread mutex and mutex attributes types for the
HIPPEROS implementation.
Enabled missing feature macros: mutex types, process sharing and timed
POSIX functions.

### CODE REVIEW

File: newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h [hunk 1]
```diff
@@ -115,9 +124,35 @@ typedef struct {
...
+#if defined(_UNIX98_THREAD_MUTEX_ATTRIBUTES)
+#define _PTHREAD_MUTEX_INITIALIZER     \
+    ((pthread_mutex_t){                \
+        .type = PTHREAD_MUTEX_DEFAULT, \
+        .is_initialized = 1,           \
+        .owner = 0xFFFFFFFFu,          \
+        .lock_count = 0u,              \
+    })
+#else
+#define _PTHREAD_MUTEX_INITIALIZER     \
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h:146]: The `else  is useless, right?

File: newlib/libc/include/sys/features.h [hunk 2]
```diff
@@ -373,6 +373,8 @@ extern "C" {
 
 #ifdef __hipperos__
 #define _POSIX_THREADS 1
```
Comment (VladimirSvoboda) [inline newlib/libc/include/sys/features.h]: Could you comment these?

File: newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h [hunk 3]
```diff
@@ -87,19 +91,24 @@ typedef struct {
...
  * one of the other mutex types.
  */
-#define PTHREAD_MUTEX_DEFAULT 3
+#define PTHREAD_MUTEX_DEFAULT PTHREAD_MUTEX_NORMAL
 
 #endif /* !defined(_UNIX98_THREAD_MUTEX_ATTRIBUTES) */
 
-#if defined(__XMK__)
-typedef unsigned int pthread_mutex_t; /* identify a mutex */
+/** Definition of the sem_t type for HIPPEROS. */
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h:98]: What was the `__XMK__`?

File: newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h [hunk 4]
```diff
@@ -87,19 +91,24 @@ typedef struct {
...
+typedef uintptr_t __sem_t;
 
+/** Mutex identifier. */
 typedef struct {
+    /** Underlying semaphore. */
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h]: Order the struct by the size of the attributes. Biggest attributes first.

**Review by VladimirSvoboda — CHANGES_REQUESTED** [review]
Small changes needed.

File: newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h (hunk 3 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h:98]: Another OS.

File: newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h (hunk 1 above)
Comment (odesenfans) [reply newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h:146]: For the current implementation, yes. But as we have ifdefs for the mutex types above, this makes sense.

File: newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h [hunk 5]
```diff
@@ -102,27 +102,36 @@ typedef uintptr_t __sem_t;
 typedef struct {
     /** Underlying semaphore. */
     __sem_t sem;
-    int is_initialized;
-    int type;
-    #if defined(_UNIX98_THREAD_MUTEX_ATTRIBUTES)
+#if defined(_UNIX98_THREAD_MUTEX_ATTRIBUTES)
+    /** Thread identifier of the owner of the mutex. */
     pthread_t owner;
```
Comment (VladimirSvoboda) [inline newlib/libc/sys/hipperos/include/sys/_pthreadtypes.h]: size_t is bigger than pthread_t

**Review by VladimirSvoboda — APPROVED** [review]

**Pull request merged**